pkg/storage/swift/swift.go

```diff
@@ -1,84 +1,294 @@
+	err = yaml.Unmarshal(content, &clouds)
+	if err != nil {
+		return emptyCloud, fmt.Errorf("failed to unmarshal clouds credentials stored in secret %v: %v", secretName, err)
+	}
+
+	return clouds.Clouds[cloudName], nil
 }
 
+// getSwiftClient returns a client that allows us to interact with the OpenStack Swift service
+// If CloudsSecret and CloudsName parameters are provided in the operator's config, then it will
+// try to read cloud config from the related secret and use it, overwriting all other parameters
+// in the operator's config (Domain, Tenant, AuthURL, etc.).
+// If those two parameters are omitted, then credentials will be taken from the operator's native
+// secret.
+func (d *driver) getSwiftClient(cr *imageregistryv1.Config) (*gophercloud.ServiceClient, error) {
+	clientOpts := new(clientconfig.ClientOpts)
+	var opts *gophercloud.AuthOptions
+
+	if cr.Spec.Storage.Swift.CloudsSecret != nil {
+		var err error
+		// If CloudsSecret and CloudsName parameters are provided then read config from clouds.yaml
+		cloud, err := d.getCloudFromSecret(cr.Spec.Storage.Swift.CloudsSecret.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		if cloud.AuthInfo != nil {
+			clientOpts.AuthInfo = cloud.AuthInfo
+			clientOpts.AuthType = cloud.AuthType
+			clientOpts.Cloud = cloud.Cloud
+			clientOpts.RegionName = cloud.RegionName
+		}
+
+		opts, err = clientconfig.AuthOptions(clientOpts)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// Otherwise take credentials from the native secret
+		cfg, err := clusterconfig.GetSwiftConfig(d.Listers)
+		if err != nil {
+			return nil, err
+		}
+
+		opts = &gophercloud.AuthOptions{
+			IdentityEndpoint: cr.Spec.Storage.Swift.AuthURL,
+			Username:         cfg.Storage.Swift.Username,
+			Password:         cfg.Storage.Swift.Password,
+			DomainID:         cr.Spec.Storage.Swift.DomainID,
+			DomainName:       cr.Spec.Storage.Swift.Domain,
+			TenantID:         cr.Spec.Storage.Swift.TenantID,
+			TenantName:       cr.Spec.Storage.Swift.Tenant,
+		}
+	}
+
+	opts.AllowReauth = true
+
+	provider, err := openstack.AuthenticatedClient(*opts)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := openstack.NewContainerV1(provider, gophercloud.EndpointOpts{})
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+// NewDriver creates new Swift driver for the Image Registry
 func NewDriver(c *imageregistryv1.ImageRegistryConfigStorageSwift, listers *regopclient.Listers) *driver {
+	var secretLister kubelisters.SecretNamespaceLister
+	if c.CloudsSecret != nil {
+		kubeconfig, err := regopclient.GetConfig()
+		if err != nil {
+			// TODO(Fedosin): an error should also be returned in this case
+			return nil
+		}
+		kubeClient, err := kubeset.NewForConfig(kubeconfig)
+		if err != nil {
+			// TODO(Fedosin): ditto
+			return nil
+		}
+		kubeInformerFactory := kubeinformers.NewSharedInformerFactoryWithOptions(kubeClient, 10*time.Minute, kubeinformers.WithNamespace(c.CloudsSecret.Namespace))
```

Comment:
In general you also need to start your factory. But I'm against of creating a new factory here. We create a new driver instance on every event from the master API, so we expect this operation to be cheap. Having new informers aren't cheap (+1 watcher, +1 store).